### PR TITLE
Speed up log value filtering in TimeSeriesProperty.

### PR DESCRIPTION
**Pull request by bolotovskyr**
#### Description of work

While doing profiling for #36069, it was discovered that time filtering of log values was one of the performance bottlenecks. This PR refactors `TimeSeriesProperty<TYPE>::createFilteredData()` as follows:

- If all time values in the `TimeSeriesProperty` object are the same, copy only the first value.
- Iterate straight over the `TimeROI` implementation, i.e. a vector of time boundaries, instead of iterating over the `TimeROI` time intervals. Generating time intervals every time is very inefficient.
- Concurrently advance the `TimeROI` vector (length N) iterator and the time values vector (length M) iterator. This provides a time efficiency of O(N+M).
- Use const and references whenever possible to facilitate compiler optimization and avoid unnecessary copying of variables.

While the above work was being done, it was noted that unlike event filtering, the `TimeROI` regions in this case should be treated as `[inclusive,inclusive]`. If they were treated as [inclusive, exclusive), the existing `HYSPECReductionTest` system test would fail. However, no unit tests in `TimeSeriesPropertyTest` were testing that "inclusive right-boundary" condition. Correspondingly, a new unit test was added to check that functionality.

#### To test:

`ctest -R KernelTest_TimeSeriesPropertyTest`

_There is no associated issue._

_This does not require release notes_ because it will never be directly observed by a user.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 1]
```diff
@@ -308,36 +311,63 @@ void TimeSeriesProperty<TYPE>::createFilteredData(const TimeROI &timeROI,
...
+      // Now try finding the first value past the end of the current ROI "use" region.
+      while (itValue != itValueEnd && itValue->time() <= *(std::next(itROI)))
+        itValue++;
+      // Include the current value, therefore, advance itEndUseValue, because std::copy works as [begin,end).
+      itEndUseValue = itValue == itValueEnd ? itValue : std::next(itValue);
+    }
+    // If we are at the last ROI "use" region or the value is not past the beginning of the next ROI "use" region, keep
+    // it for the current ROI "use" region.
+    else if (std::distance(itROI, itROIEnd) == 2 ||
+             (std::distance(itROI, itROIEnd) > 2 && itValue->time() < *(std::next(itROI, 2)))) {
```
Comment (jmborr) [inline Framework/Kernel/src/TimeSeriesProperty.cpp:353]: Is `(std::distance(itROI, itROIEnd) > 2` necessary? if the loop works as intended this seems redundant.

**Review by jmborr — CHANGES_REQUESTED** [review]
In `TimeSeriesProperty<TYPE>::removeDataOutsideTimeROI`, why are we using:
```
std::vector<TimeValueUnit<TYPE>> mp_copy;
  createFilteredData(timeROI, mp_copy);
```  
Instead of
```
  m_values.clear();
  createFilteredData(timeROI, m_values);
```
?

File: Framework/Kernel/src/TimeSeriesProperty.cpp (hunk 1 above)
Comment (bolotovskyr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:353]: It is correct. We're advancing `itROI` **only** as long as `*(std::next(itROI, 2)) <= itValue->time()` is true. However, that condition **can't be checked** unless `std::distance(itROI, itROIEnd) > 2`

Comment (jmborr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:366]: I don't understand your explanation :disappointed:  
If `std::distance(itROI, itROIEnd)` is not `2`, then it must be either `0` or `> 2`. The case `0` you dealt with before, so that leaves `std::distance(itROI, itROIEnd) > 2` being true. Hence, evaluating condition `(std::distance(itROI, itROIEnd) > 2` is a moot point.
Am I missing something here?

Comment (bolotovskyr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:366]: Could you better explain to me what you mean by "before" in:

> The case 0 you dealt with before, so that leaves std::distance(itROI, itROIEnd) > 2 being true

Are you referring to L336:
`  while (itROI != itROIEnd && itValue != itValueEnd) {
`
If you do, that does not apply here, because it's not the same loop.

Comment (jmborr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:366]: When loop `while (itROI != itROIEnd && itValue != itValueEnd)` starts executing, we necessarily have `std::distance(itROI, itROIEnd) >= 2`, so at this point we have dealt with condition `std::distance(itROI, itROIEnd) == 0`.
However, I just noticed now the inner `while` loop:
```c++
    while (std::distance(itROI, itROIEnd) > 2 && *(std::next(itROI, 2)) <= itValue->time())
      std::advance(itROI, 2);
```
If this loops brings `itROI` to `itROIEnd`, then we'll have again `std::distance(itROI, itROIEnd) == 0` and in this case we indeed need to check for condition `std::distance(itROI, itROIEnd) > 2`. So, is this inner `while` loop the reason for explicitly introducing condition `std::distance(itROI, itROIEnd) > 2`?

Comment (bolotovskyr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:366]: Yes, that is the reason.

Comment (jmborr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:366]: OK, then. This conversation's resolved.
Sorry it took me so long to find out about the inner loop. I guess I wasted your time

Comment (bolotovskyr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:366]: No, you haven't.  I agree with you, when the inner loop just starts, the distance check is redundant.  However, it's not obvious how to avoid that extra check without introducing even more complexity to the code.

BTW, if you take a look at the next statement, which is a "hidden loop"
```
itValue = std::lower_bound(itValue, itValueEnd, *itROI,
                               [](const auto &value, const auto &roi_time) { return value.time() < roi_time; });
```
there is basically the same inefficiency there, because `std::lower_bound` implementation must exit if `itValue` is `itValueEnd`, but the first check for that is redundant, because we have already checked it in our outer loop!

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 2]
```diff
@@ -308,36 +311,63 @@ void TimeSeriesProperty<TYPE>::createFilteredData(const TimeROI &timeROI,
     filteredData.push_back(m_values.front());
```
[the remark this answers is not in the document]
Comment (bolotovskyr) [reply Framework/Kernel/src/TimeSeriesProperty.cpp:321]: This is pre-existing logic. According to Pete, "A TimeSeriesProperty with no values doesn't make any sense, and deleting a log altogether would surprise people."

**Review by jmborr — APPROVED** [review]

**Comment by bolotovskyr** [comment]
WRT:

> In TimeSeriesProperty<TYPE>::removeDataOutsideTimeROI, why are we using:
> 
> std::vector<TimeValueUnit<TYPE>> mp_copy;
>   createFilteredData(timeROI, mp_copy);
> 
> Instead of
> 
>   m_values.clear();
>   createFilteredData(timeROI, m_values);
> 
> ?

`createFilteredData` creates a new vector of values out of existing `m_values`. If we cleared `m_values` before calling `createFilteredData`, we would have no values to choose from.  Thanks to @peterfpeterson for helping with this question.
